calyx/src/passes/merge_static_par.rs

```diff
@@ -0,0 +1,121 @@
+#[derive(Default)]
+pub struct MergeStaticPar;
+
+impl Named for MergeStaticPar {
+    fn name() -> &'static str {
+        "merge-static-par"
+    }
+
+    fn description() -> &'static str {
+        "merge static pars when they have the same static time"
+    }
+}
+
+impl Visitor for MergeStaticPar {
+    fn finish_par(
+        &mut self,
+        s: &mut ir::Par,
+        comp: &mut ir::Component,
+        sigs: &ir::LibrarySignatures,
+        _comps: &[ir::Component],
+    ) -> VisResult {
+        let mut static_group: LinkedHashMap<u64, Vec<RRC<ir::Group>>> =
+            LinkedHashMap::new();
+        let (e_stmts, n_stmts): (Vec<ir::Control>, Vec<ir::Control>) =
+            mem::take(&mut s.stmts)
+                .into_iter()
+                .partition(|stmt| matches!(stmt, ir::Control::Enable(_)));
+
+        s.stmts.extend(n_stmts);
+
+        for stmt in e_stmts {
+            if let ir::Control::Enable(data) = stmt {
+                let group = &data.group;
+                let static_time: u64 =
+                    *group.borrow().attributes.get("static").unwrap();
+                if !static_group.contains_key(&static_time) {
+                    static_group.insert(static_time, Vec::new());
+                }
+                static_group
+                    .get_mut(&static_time)
+                    .unwrap()
+                    .push(Rc::clone(group));
+            }
+        }
+
+        for (key, value) in static_group {
+            if value.len() != 1 {
+                let mut builder = ir::Builder::new(comp, sigs);
+                let grp = builder.add_group("msp");
+                let mut assignments: Vec<ir::Assignment> = Vec::new();
+                for group in value.iter() {
+                    assignments.extend(group.borrow().assignments.clone());
+                }
+
+                let (done_asmts, asmts): (
+                    Vec<ir::Assignment>,
+                    Vec<ir::Assignment>,
+                ) = mem::take(&mut assignments)
+                    .into_iter()
+                    .partition(|x| x.dst.borrow().is_hole());
+
+                grp.borrow_mut().assignments.extend(asmts);
+
+                let mut ports: Vec<ir::Guard> = Vec::new();
+                for asmt in done_asmts.clone() {
+                    let grd: ir::Guard = ir::Guard::Port(asmt.src);
+                    ports.push(grd);
+                    ports.push(*asmt.guard);
+                }
+
+                let mut fin_grd: ir::Guard = ir::Guard::True;
+                for grd in ports {
+                    fin_grd &= grd;
+                }
+
+                let cst = builder.add_constant(1, 1);
+
+                let done_asmt = builder.build_assignment(
+                    grp.borrow().get("done"),
+                    cst.borrow().get("out"),
+                    fin_grd,
+                );
+
+                {
+                    grp.borrow_mut().assignments.push(done_asmt);
+                }
```

Comment:
Why does this need to be indented? What error shows up if you don't do it?

Reply:
I thought because I used borrow_mut() again before the end of the current lexical scope, that it was going to cause a panic, but in fact it did not so I deleted the curly braces.